rep-2011.rst

```diff
@@ -0,0 +1,1314 @@
+
+- because it's on another computer, or
+- because it is from a different distribution of ROS, or
+- because it was built in a different workspace, or
+- because the application has not been restarted since recompiling a change to the type being used
+
+In any case, it is useful to have a mechanism to convey the type descriptions from the source of the data to other nodes, which we describe here as "type description distribution".
+
+Furthermore, this feature should be agnostic to the underlying middleware and serialization library, as two endpoints may not have the same rmw implementation, or the data may have been serialized to a different format in the case of playback of a recording.
+
+Sending the Type Description
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Type descriptions will be provided by a ROS Service called ``~/_get_type_description``, which will be offered by each node.
+There will be a single ROS Service per node, regardless of the number of publishers or subscriptions on that node.
+
+This service may be optional, for example being enabled or disabled when creating the ROS Node.
+
+.. TODO::
+
+    How can we detect when a remote node is not offering this service?
+    It's difficult to differentiate between "the Service has not been created yet, but will be" and "the Service will never be created".
+    Should we use a ROS Parameter to indicate this?
+    But then what if remote access to Parameters (another optional Service) is disabled?
+    Perhaps we need a "services offered" list which is part of the Node metadata, which is sent for each node in the rmw implementation, but that's out of scope for this REP.
+
+A service request to this ROS Service will comprise of the type name and the type version hash, which is distributed during discovery of endpoints and will be accessible through the ROS Graph API, as described in previous sections.
+The ROS Service server will respond with the type description and any necessary metadata needed to do Run-Time Interface Reflection.
+This service is not expected to be called frequently, and is likely to only occur when new topic or service endpoints are created, and even then, only if the endpoint type hashes do not match.
+
+.. TODO::
+
+    Should each endpoint be asked about their type/version pair or should we assume that the type/version pair guarantees a unique type description and therefore reuse past queries?
+    (wjwwood) I am leaning towards assuming the type name and type version hash combo as being a unique identifier.
+
+Type Description Contents and Format
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The response sent by the ROS Service server will contain a combination of the original ``idl`` or ``msg`` file's content, as well as any necessary information to serialize and deserialize the raw message buffers sent on the topic.
```

Comment:
It's probably an important detail that the full `idl` or `msg` definition of all field types needs to be provided along with the toplevel definition, otherwise there may be an unresolved incompatibility between types at a field level. It's worth noting that ROS1 tries to solve this by concatenating all of its `msg` definitions using `gendeps --cat` and sending this in the connection header for new connections.

Reply:
Yup! You're right! They're meant to be sent along the main type description under the `referenced_type_descriptions` field.

At least, for any non-primitive types.

Do you see value in concatenating the referenced file contents as well (including the comments)? Or would a parsed definition suffice?

Reply:
> Do you see value in concatenating the referenced file contents as well (including the comments)? Or would a parsed definition suffice?

It depends on how frozen-in-time the parsing procedure is. If the parsed definition evolves over time, I'd keep the original file content around, because then if there's a Type Description schema mismatch, a node can fall back to rebuilding the type description from the original file content. Stripping comments seems safe enough, since at least then the content is still valid IDL or MSG.

Reply:
@wjwwood regarding storing a type definition (msg or idl) as a string, I wanted to call your attention to work that has been happening over in [mcap](https://mcap.dev/).

The specific problem I am referencing is how to store _both_ the message definition and all direct and transitive dependencies in the proposed `type_description_raw` field. ROS 1 solves this using the [`gendeps --cat`](http://wiki.ros.org/roslib/gentools) format (mentioned by @james-rms above), which consists of msg files delimited by 80 `=` characters plus a message name line.

For storing concatenated source message definitions in MCAP, we opted to follow the ROS 1 convention for msg files, which we use for both ROS 1 and ROS 2 data. For IDL source message definitions, we modified the concatenation slightly so that all types including the top level type are explicitly named.

You can see our specification for this [here](https://github.com/foxglove/mcap/blob/main/docs/specification/appendix.md#ros2msg), and discussion on the spec PR at https://github.com/foxglove/mcap/pull/553. This spec was then implemented in https://github.com/ros-tooling/rosbag2_storage_mcap/pull/53.

I think it would be best if we can use this same concatenation standard for REP 2011 types.

Reply:
It was pointed out to me that you could avoid concatenating message definition strings entirely by providing an array of types (the top level type and any dependent types, each with a name, format, and string body).